zypp-plugin/snapper_zypp_plugin.cc

```diff
@@ -0,0 +1,349 @@
+// getenv
+#include <stdlib.h>
+// getppid
+#include <sys/types.h>
+#include <unistd.h>
+// fnmatch
+#include <fnmatch.h>
+
+#include <boost/regex.hpp>
+#include <boost/format.hpp>
+using boost::format;
+
+#include <iostream>
+#include <map>
+#include <string>
+using namespace std;
+
+#include <rapidjson/document.h>
+
+#include "dbus/DBusConnection.h"
+#include "snapper/Exception.h"
+using snapper::Exception;
+using snapper::CodeLocation;
+#include "client/commands.h"
+#include "client/errors.h"
+#include "snapper/Log.h"
+#include "snapper/XmlFile.h"
+
+#include "zypp_commit_plugin.h"
+
+ostream& operator <<(ostream& os, set<string> ss) {
+    bool seen_first = false;
+    os << '{';
+    for(auto s: ss) {
+	if (seen_first) {
+	    os << ", ";
+	}
+	os << s;
+	seen_first = true;
+    }
+    os << '}';
+    return os;
+}
+
+class ProgramOptions {
+public:
+    string plugin_config;
+    string snapper_config;
+    DBusBusType bus;
+
+    ProgramOptions()
+    : plugin_config("/etc/snapper/zypp-plugin.conf")
+    , snapper_config("root")
+    , bus(DBUS_BUS_SYSTEM)
+    {
+	const char * s;
+
+	s = getenv("SNAPPER_ZYPP_PLUGIN_CONFIG");
+	if (s != nullptr) {
+	    plugin_config = s;
+	}
+
+	s = getenv("SNAPPER_ZYPP_PLUGIN_SNAPPER_CONFIG");
+	if (s != nullptr) {
+	    snapper_config = s;
+	}
+
+	s = getenv("SNAPPER_ZYPP_PLUGIN_DBUS_SESSION");
+	if (s != nullptr) {
+            bus = DBUS_BUS_SESSION;
+        }
+    }
+};
+
+class SnapperZyppPlugin : public ZyppCommitPlugin {
+public:
+    struct SolvableMatcher {
+	enum class Kind {
+	    Glob,
+	    Regex
+	};
+	string pattern;
+	Kind kind;
+	bool important;
+
+	SolvableMatcher(string apattern, Kind akind, bool aimportant)
+	: pattern(apattern)
+	, kind(akind)
+	, important(aimportant)
+	{}
+
+	bool match(const string& solvable) {
+	    y2deb("match? " << solvable << " by " << ((kind == Kind::Glob)? "GLOB '": "REGEX '") << pattern << '\'');
+	    bool res;
+	    if (kind == Kind::Glob) {
+		static const int flags = 0;
+		res = fnmatch(pattern.c_str(), solvable.c_str(), flags) == 0;
+	    }
+	    else {
+		// POSIX Extended Regular Expression Syntax
+		boost::regex rx_pattern(pattern, boost::regex::extended);
```

Comment:
regex compilation can fail (which would be a user error). Some error logging seems appropriate.

Reply:
Example from zypp-plugin/testsuite/3-invalid-xml.test,
This will appear in zypper.log:
```console
DEBUG:match? mypkg by REGEX 'an invalid [ regex'
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::regex_error> >'
  what():  Unmatched [ or [^ in character class declaration.  The error occurred while parsing the regular expression fragment: 'id [ regex>>>HERE>>>'.
```

Reply:
Now slightly changed after kicking out boost_regex:
```console
DEBUG:match? mypkg by REGEX 'an invalid [ regex'
terminate called after throwing an instance of 'std::runtime_error'
  what():  Regex compilation error: Unmatched [ or [^
```

Reply:
So I assume this is OK.